src/test/regress/citus_tests/common.py

```diff
@@ -315,8 +350,817 @@ def eprint(*args, **kwargs):
     print(*args, file=sys.stderr, **kwargs)
 
 
-def run(command, *args, shell=True, **kwargs):
+def run(command, *args, check=True, shell=True, silent=False, **kwargs):
     """run runs the given command and prints it to stderr"""
 
-    eprint(f"+ {command} ")
-    return subprocess.run(command, *args, check=True, shell=shell, **kwargs)
+    if not silent:
+        eprint(f"+ {command} ")
+    if silent:
+        kwargs.setdefault("stdout", subprocess.DEVNULL)
+    return subprocess.run(command, *args, check=check, shell=shell, **kwargs)
+
+
+def sudo(command, *args, shell=True, **kwargs):
+    """
+    A version of run that prefixes the command with sudo when the process is
+    not already run as root
+    """
+    effective_user_id = os.geteuid()
+    if effective_user_id == 0:
+        return run(command, *args, shell=shell, **kwargs)
+    if shell:
+        return run(f"sudo {command}", *args, shell=shell, **kwargs)
+    else:
+        return run(["sudo", *command])
+
+
+# this is out of ephemeral port range for many systems hence
+# it is a lower chance that it will conflict with "in-use" ports
+PORT_LOWER_BOUND = 10200
+
+# ephemeral port start on many Linux systems
+PORT_UPPER_BOUND = 32768
+
+next_port = PORT_LOWER_BOUND
+
+
+def cleanup_test_leftovers(nodes):
+    for node in nodes:
+        node.cleanup_subscriptions()
+
+    for node in nodes:
+        node.cleanup_publications()
+
+    for node in nodes:
+        node.cleanup_logical_replication_slots()
+
+    for node in nodes:
+        node.cleanup_schemas()
+
+    for node in nodes:
+        node.cleanup_users()
+
+
+class PortLock:
+    """PortLock allows you to take a lock an a specific port.
+
+    While a port is locked by one process, other processes using PortLock won't
+    get the same port.
+    """
+
+    def __init__(self):
+        global next_port
+        while True:
+            next_port += 1
+            if next_port >= PORT_UPPER_BOUND:
+                next_port = PORT_LOWER_BOUND
+
+            self.lock = filelock.FileLock(Path(gettempdir()) / f"port-{next_port}.lock")
+            try:
+                self.lock.acquire(timeout=0)
+            except filelock.Timeout:
+                continue
+
+            if FORCE_PORTS:
+                self.port = next_port
+                break
+
+            with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+                try:
+                    s.bind(("127.0.0.1", next_port))
+                    self.port = next_port
+                    break
+                except Exception:
+                    continue
+
+    def release(self):
```

Comment:
Can we release the lock at object's destruction hook instead of caller is responsible to release it?

Reply:
We could but destruction hooks have their own problems: https://stackoverflow.com/a/1481512/2570866

One thing is that the hook might never be called. So, I'd rather keep calling `release` explicit.

Reply:
Yes, I use context managers in a few other places in this PR. But at the call site where PortLock is used it's not possible because the open and close are in separate function calls. I could ofcourse still implement the context manager interface, but if it isn't used anyway it seems a bit useless.